tiledb/sm/tile/generic_tile_io.h

```diff
@@ -0,0 +1,216 @@
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class GenericTileIO.
+ */
+
+#ifndef TILEDB_GENERIC_TILE_IO_H
+#define TILEDB_GENERIC_TILE_IO_H
+
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/enums/encryption_type.h"
+#include "tiledb/sm/filter/filter_pipeline.h"
+#include "tiledb/sm/misc/uri.h"
+
+namespace tiledb {
+namespace sm {
+
+class StorageManager;
+class Tile;
+
+/** Handles IO (reading/writing) for tiles. */
+class GenericTileIO {
+ public:
+  /* ********************************* */
+  /*       PUBLIC TYPE DEFINITIONS     */
+  /* ********************************* */
+
+  /**
+   * Header information for a generic tile. A generic tile is a tile residing
+   * together with its metadata in one contiguous byte region of a file. This is
+   * as opposed to a regular tile, where the metadata resides separately from
+   * the tile data itself.
+   */
+  struct GenericTileHeader {
+    /** Size in bytes of the non-filters part of the serialized header. */
+    static const uint64_t BASE_SIZE =
+        3 * sizeof(uint64_t) + 2 * sizeof(char) + 2 * sizeof(uint32_t);
+    /** Format version number of the tile. */
+    uint32_t version_number;
+    /** Persisted (e.g. compressed) size of the tile. */
+    uint64_t persisted_size;
+    /** Uncompressed size of the tile. */
+    uint64_t tile_size;
+    /** Datatype of the tile. */
+    uint8_t datatype;
+    /** Cell size of the tile. */
+    uint64_t cell_size;
+    /** The type of the encryption used in filtering. */
+    uint8_t encryption_type;
+    /** Number of bytes in the serialized filter pipeline instance. */
+    uint32_t filter_pipeline_size;
+    /** Filter pipeline used to filter the tile. */
+    FilterPipeline filters;
+
+    /** Constructor. */
+    GenericTileHeader()
+        : version_number(constants::format_version)
+        , persisted_size(0)
+        , tile_size(0)
+        , datatype((uint8_t)Datatype::ANY)
+        , cell_size(0)
+        , encryption_type((uint8_t)EncryptionType::NO_ENCRYPTION)
+        , filter_pipeline_size(0) {
+    }
+  };
+
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /**
+   * Constructor.
+   *
+   * @param storage_manager The storage manager.
+   * @param uri The name of the file that stores data.
+   */
+  GenericTileIO(StorageManager* storage_manager, const URI& uri);
+
+  GenericTileIO() = delete;
+  GenericTileIO(const GenericTileIO& other) = delete;
```

Comment:
This is work item #4 on the ticket:
Deleting the copy/move constructor/operators are a common enough technique that we have macros for them here: https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/misc/macros.h

Replace lines 106-109 here with:
```
DISABLE_COPY_AND_COPY_ASSIGN(GenericTileIO);
DISABLE_MOVE_AND_MOVE_ASSIGN(GenericTileIO);
```